src/ws/hooks/app.ts

```diff
@@ -0,0 +1,180 @@
+          }
+        };
+        websocketClient.subscribe(channel, handler);
+
+        return function cleanup() {
+          websocketClient.unsubscribe(channel, handler);
+        };
+      }, [itemId]);
+    },
+    /**
+     * React hook to subscribe to the updates of the app data for
+     * the given item ID.
+     * @param itemId The ID of the item of which to observe updates
+     */
+    useAppActionsUpdates: (itemId?: UUID | null) => {
+      const queryClient = useQueryClient();
+      useEffect(() => {
+        if (!itemId || itemId?.length === 0) {
+          return () => {
+            // do nothing
+            console.warn('No correct itemId provided.');
+          };
+        }
+
+        const channel: Channel = { name: itemId, topic: APP_ACTIONS_TOPIC };
+        const appActionKey = buildAppActionsKey(itemId);
+
+        const handler = (event: AppActionEvent) => {
+          if (event.kind === 'app-actions') {
+            const appDataList: List<AppActionRecord> | undefined =
+              queryClient.getQueryData(appActionKey);
+            const newAppAction: AppActionRecord = convertJs(event.appAction);
+            switch (event.op) {
+              case 'post': {
+                if (appDataList?.findIndex(({ id }) => id === newAppAction.id) === -1)
+                  queryClient.setQueryData(appActionKey, appDataList?.push(newAppAction));
+                break;
+              }
+              default:
+                console.warn('unhandled event for useAppActionsUpdates');
+                break;
+            }
+          }
+        };
+        websocketClient.subscribe(channel, handler);
+
+        return function cleanup() {
+          websocketClient.unsubscribe(channel, handler);
+        };
+      }, [itemId]);
+    },
+    /**
+     * React hook to subscribe to the updates of the app data for
+     * the given item ID.
+     * @param itemId The ID of the item of which to observe updates
+     */
+    useAppSettingsUpdates: (itemId?: UUID | null) => {
+      const queryClient = useQueryClient();
+      useEffect(() => {
+        if (!itemId || itemId?.length === 0) {
+          return () => {
+            // do nothing
+            console.warn('No correct itemId provided.');
+          };
+        }
+
+        const channel: Channel = { name: itemId, topic: APP_SETTINGS_TOPIC };
+        const appSettingsKey = buildAppSettingsKey(itemId);
+
+        const handler = (event: AppSettingEvent) => {
+          if (event.kind === 'app-settings') {
+            const appSettingList: List<AppSettingRecord> | undefined =
+              queryClient.getQueryData(appSettingsKey);
+            const newAppSetting: AppSettingRecord = convertJs(event.appSetting);
+            switch (event.op) {
+              case 'post': {
+                if (appSettingList?.findIndex(({ id }) => id === newAppSetting.id) === -1)
+                  queryClient.setQueryData(appSettingsKey, appSettingList?.push(newAppSetting));
```

Comment:
Use some instead of findIndex, consider the null / undefined case. Also add curly braces around the if block to avoid the goto bug pitfall.